src/storage-client/src/controller/command_wals.rs

```diff
@@ -0,0 +1,154 @@
+//! or write them. This identifies shards we no longer use, but did not have the
+//! opportunity to finalize before e.g. crashing.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use differential_dataflow::lattice::Lattice;
+use timely::order::TotalOrder;
+use timely::progress::Timestamp;
+
+use mz_ore::now::EpochMillis;
+use mz_persist_client::ShardId;
+use mz_persist_types::Codec64;
+use mz_proto::RustType;
+use mz_repr::TimestampManipulation;
+use mz_stash::{self, TypedCollection};
+
+use crate::client::{StorageCommand, StorageResponse};
+use crate::controller::{ProtoStorageCommand, ProtoStorageResponse};
+
+use super::Controller;
+
+pub(super) static SHARD_FINALIZATION: TypedCollection<ShardId, ()> =
+    TypedCollection::new("storage-shards-to-finalize");
+
+impl<T> Controller<T>
+where
+    T: Timestamp + Lattice + TotalOrder + Codec64 + From<EpochMillis> + TimestampManipulation,
+    StorageCommand<T>: RustType<ProtoStorageCommand>,
+    StorageResponse<T>: RustType<ProtoStorageResponse>,
+
+    Self: super::StorageController<Timestamp = T>,
+{
+    /// `true` if shard is in register for shards marked for finalization.
+    pub(super) async fn shards_registered_for_finalization(&mut self, shard: ShardId) -> bool {
+        SHARD_FINALIZATION
+            .peek_key_one(&mut self.state.stash, shard)
+            .await
+            .expect("must be able to connect to stash")
+            .is_some()
+    }
+
+    /// Register shards for finalization. This must be called if you intend to
+    /// finalize shards, before you perform any work to e.g. replace one shard
+    /// with another.
+    ///
+    /// The reasoning behind this is that we need to identify the intent to
+    /// finalize a shard so we can perform the finalization on reboot if we
+    /// crash and do not find the shard in use in any collection.
+    #[allow(dead_code)]
+    pub(super) async fn register_shards_for_finalization<I>(&mut self, entries: I)
+    where
+        I: IntoIterator<Item = ShardId>,
+    {
+        SHARD_FINALIZATION
+            .insert_without_overwrite(
+                &mut self.state.stash,
+                entries.into_iter().map(|key| (key, ())),
+            )
+            .await
+            .expect("must be able to write to stash")
+    }
+
+    /// Removes the shard from the finalization register.
+    ///
+    /// This is appropriate to do if you can guarantee that the shard has been
+    /// finalized or find the shard is still in use by some collection.
+    pub(super) async fn clear_from_shard_finalization_register(
+        &mut self,
+        shards: &BTreeSet<ShardId>,
+    ) {
+        SHARD_FINALIZATION
+            .delete(&mut self.state.stash, |k, _v| shards.contains(k))
+            .await
+            .expect("must be able to write to stash")
+    }
+
+    /// Reconcile the state of `SHARD_FINALIZATION_WAL` with
+    /// `super::METADATA_COLLECTION` on boot.
+    pub(super) async fn reconcile_shards(&mut self) {
```

Comment:
So in the expected case we do one stash operation, see that there's nothing to do and return. In the more involved case we do more stash operations, but we expect to rarely hit that?

Reply:
Correct; we would find this only in the case that we crashed during `rewrite_collection_metadata`.